src/crimson/os/seastore/onode_manager/simple-fltree/onode_node.h

```diff
@@ -0,0 +1,1339 @@
+      assert(0);
+    } else if constexpr (item_in_key) {
+      keys[slot].update(oid);
+    } else {
+      const auto& [prefix, suffix] = key_at(slot);
+      int16_t delta = partial_key_t::size_from(oid) - suffix.size();
+      if (delta > 0) {
+        // shift the cells sitting at its left side
+        auto first = keys[slot].offset;
+        auto last = keys[count - 1].offset;
+        std::memmove(from_end(last + delta), from_end(last), last - first);
+        // update the pointers
+        for (key_t* key = keys + slot; key < keys + count; key++) {
+          key->offset += delta;
+        }
+      }
+      keys[slot].update(oid);
+      auto p = from_end(keys[slot].offset);
+      auto partial_key = reinterpret_cast<partial_key_t*>(p);
+      partial_key->set(oid);
+      // we don't update item here
+    }
+  }
+
+  // @return the number of element to grab
+  std::pair<unsigned, uint16_t> calc_grab_front(uint16_t min_grab, uint16_t max_grab) const {
+    // TODO: split by likeness
+    uint16_t grabbed = 0;
+    uint16_t used = used_space();
+    int n = 0;
+    for (; n < count; n++) {
+      const auto& [prefix, suffix] = key_at(n);
+      uint16_t to_grab = sizeof(prefix) + size_of(suffix);
+      if constexpr (!item_in_key) {
+        const auto& item = item_at(prefix);
+        to_grab += size_of(item);
+      }
+      if (grabbed + to_grab > max_grab) {
+        break;
+      }
+      grabbed += to_grab;
+    }
+    if (grabbed >= min_grab) {
+      if (n == count) {
+        return {n, grabbed};
+      } else if (!is_underflow(used - grabbed)) {
+        return {n, grabbed};
+      }
+    }
+    return {0, 0};
+  }
+
+  std::pair<unsigned, uint16_t> calc_grab_back(uint16_t min_grab, uint16_t max_grab) const {
+    // TODO: split by likeness
+    uint16_t grabbed = 0;
+    uint16_t used = used_space();
+    for (int i = count - 1; i >= 0; i--) {
+      const auto& [prefix, suffix] = key_at(i);
+      uint16_t to_grab = sizeof(prefix) + size_of(suffix);
+      if constexpr (!item_in_key) {
+        const auto& item = item_at(prefix);
+        to_grab += size_of(item);
+      }
+      grabbed += to_grab;
+      if (is_underflow(used - grabbed)) {
+        return {0, 0};
+      } else if (grabbed > max_grab) {
+        return {0, 0};
+      } else if (grabbed >= min_grab) {
+        return {i + 1, grabbed};
+      }
+    }
+    return {0, 0};
+  }
+
+  template<int LeftN, class Mover>
+  void grab_from_left(
+    node_t<BlockSize, LeftN, NodeType>& left,
+    unsigned n, uint16_t bytes,
+    Mover& mover) {
+    // TODO: rebuild keys if moving across different layouts
+    //       group by likeness
+    shift_right(n, bytes);
+    mover.move_from(left.count - n, 0, n);
+  }
+  // have to define a set of trap functions
+  template<int LeftN, class Mover, ntype_t O = flip_ntype(NodeType)>
+  void grab_from_left(node_t<BlockSize, LeftN, O>&,
+                      unsigned, uint16_t,
+                      Mover&) {
+    assert(0);
```

Comment:
`abort()` or `ceph_abort()` instead? You want to let compiler define-out this, I think.

Reply:
i insist. as there is no need to increase the size of executable because something will never be executed. the whole idea is to please the compiler.